CLI: Implement `auth login` command

fredrikekelund opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes STU-946

## Proposed Changes

This PR is based on @youknowriad's work in #1958. It implements a `studio auth login` command that:

1. Opens a WordPress.com OAuth2 authorization URL in the browser
2. The OAuth2 authorization page redirects to a developer.wordpress.com page that renders a convenient button for copying the generated access token
3. Once copied, the access token can be pasted back to the terminal

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Check out 156890-ghe-Automattic/wpcom-a8c-themes on your sandbox
2. Sandbox `developer.wordpress.com`
3. Open Studio and log out from WordPress.com
4. Close Studio
5. Run `npm run cli:build`
6. Run `node dist/cli/main.js auth login`
7. Finish the authorization process and paste the token back to the terminal
8. Open Studio again
9. Ensure you are logged in to WordPress.com

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

fredrikekelund committed:
CLI: Implement `auth login` command

fredrikekelund commented on cli/lib/browser.ts:
```diff
@@ -0,0 +1,53 @@
+			args = [ url ];
+			break;
+		case 'win32':
+			cmd = 'PowerShell';
+			args = [
+				'-NoProfile',
+				'-NonInteractive',
+				'-ExecutionPolicy',
+				'Bypass',
+				'-Command',
+				'Start',
+				`"${ url }"`,
+			];
```
I have yet to test this. This was taken from Sindre Sorhus's [open module](https://github.com/sindresorhus/open/).

fredrikekelund committed:
Fix lint and tests

fredrikekelund commented:
Working on a test right now.

fredrikekelund committed:
Add tests, tweak command

fredrikekelund committed:
Consider expiration time

fredrikekelund committed:
Move logic around

fredrikekelund commented:
Looking at how many changes I had to make today, I see that I opened this PR a little early. In any case, it's good for a full review now 👍

fredrikekelund committed:
Unwrap

fredrikekelund committed:
Fix type error

fredrikekelund committed:
Address review comments

fredrikekelund committed:
Reject unsupported platforms

sejas commented on common/logger-actions.ts:
```diff
@@ -1,6 +1,11 @@
 
+export enum AuthCommandLoggerAction {
+	LOGIN = 'login',
+	LOGOUT = 'logout',
```
Nit: We don't really need the logout at this point.

fredrikekelund replied:
I'll move this to #2027

sejas commented on cli/commands/auth/login.ts:
```diff
@@ -0,0 +1,96 @@
+			const userData = await readAppdata();
+
+			const now = new Date();
+			const twoWeeksInSeconds = 2 * 7 * 24 * 60 * 60;
```
I just saw that the `expiresIn` in appData is `1209600` seconds or two weeks, but I always thought the token was valid for months. 🤔

fredrikekelund replied:
This is governed by the back-end. We could look into this further if we think two weeks is too short

sejas reviewed (approved):
I confirm that the login works as expected, and I was able to authenticate by copying and pasting the token. Here are a couple of suggestions to improve the experience:
- We could  avoid hiding the token so the user can see they pasted the correct value.
- We could automatically try three times when the user enters an invalid token, similarly to SSH connections. Currently the feedback is very rough `✖ Failed to fetch user info: The OAuth2 token is invalid.`, we could at least invite the user to try again.

<img width="646" height="268" alt="Image" src="https://github.com/user-attachments/assets/6c4241e6-1189-4e33-aa7a-d54bf32f7c76" />

youknowriad commented:
I like how simple this is. Good work :)

fredrikekelund committed:
Merge branch 'trunk' into f26d/cli-auth-login-command

fredrikekelund committed:
Address review feedback

fredrikekelund committed:
Fix test

fredrikekelund committed:
Merge branch 'trunk' into f26d/cli-auth-login-command

github-actions[bot] committed:
docs: Update CLI documentation for auth login command

Add comprehensive documentation for the new `studio auth login` command
including usage instructions, flow description, and important notes about
token lifetime and authentication sharing between CLI and desktop app.

Co-authored-by: Fredrik Rombach Ekelund <undefined@users.noreply.github.com>

fredrikekelund committed:
Restore package-lock.json

fredrikekelund committed:
Install inquirer dependency again

fredrikekelund committed:
Tweak

bcotrim commented on cli/commands/auth/login.ts:
```diff
@@ -0,0 +1,99 @@
+		user = await getUserInfo( accessToken );
+		logger.reportSuccess( __( 'Authentication completed successfully!' ) );
+	} catch ( error ) {
+		logger.reportError( new LoggerError( __( 'Authentication failed. Please try again.' ) ) );
```
I agree that a retry logic would improve the UX to account for a mistake paste or enter

<img width="1368" height="234" alt="Image" src="https://github.com/user-attachments/assets/de0e7b3e-a2f5-42f2-83c2-93e9b817314c" />

bcotrim replied:
Another idea would be to have a "click here if you aren't automatically redirected" to allow users to manually open the browser tab.

Neither of these are blockers in my opinion, just considerations.

bcotrim reviewed (approved):
LGTM 👍
